store/cachekvstore.go

```diff
@@ -159,29 +160,29 @@ func (ci *cacheKVStore) iterator(start, end []byte, ascending bool) Iterator {
 		parent = ci.parent.ReverseIterator(start, end)
 	}
 
-	items := ci.dirtyItems(ascending)
+	items := ci.dirtyItems(start, end, ascending)
 	cache = newMemIterator(start, end, items)
 
 	return newCacheMergeIterator(parent, cache, ascending)
 }
 
 // Constructs a slice of dirty items, to use w/ memIterator.
-func (ci *cacheKVStore) dirtyItems(ascending bool) []cmn.KVPair {
-	items := make([]cmn.KVPair, 0, len(ci.cache))
+func (ci *cacheKVStore) dirtyItems(start, end []byte, ascending bool) []cmn.KVPair {
+	items := make([]cmn.KVPair, 0)
 
 	for key, cacheValue := range ci.cache {
 		if !cacheValue.dirty {
 			continue
 		}
-
-		items = append(items, cmn.KVPair{Key: []byte(key), Value: cacheValue.value})
+		if dbm.IsKeyInDomain([]byte(key), start, end) {
```

Comment:
Nice! Was this a bug before?

Reply:
I think the location of the logic just moved.

Reply:
Before we used to push all dirty items to the memIterator, which wasn't necessary.  This is more efficient.